src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs

```diff
@@ -0,0 +1,148 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers.RouteHandlers;
+
+public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
+{
+    private static void DisallowNonParsableComplexTypesOnParameters(
+        in OperationAnalysisContext context,
+        IInvocationOperation invocation,
+        IMethodSymbol methodSymbol)
+    {
+        var routePatternArgument = invocation.Arguments[1];
+
+        if (routePatternArgument.Syntax is not ArgumentSyntax routePatternArgumentSyntax ||
+            routePatternArgumentSyntax.Expression is not LiteralExpressionSyntax routePatternArgumentLiteralSyntax)
+        {
+            return;
+        }
+
+        var virtualChars = CSharpVirtualCharService.Instance.TryConvertToVirtualChars(routePatternArgumentLiteralSyntax.Token);
+        var parser = RoutePatternParser.TryParse(virtualChars, false);
+
+        var wellKnownTypes = WellKnownTypes.GetOrCreate(context.Compilation);
+
+        foreach (var handlerDelegateParameter in methodSymbol.Parameters)
+        {
+            // If the parameter is decorated with a FromServices attribute then we can skip it.
+            var fromServicesAttributeTypeSymbol = wellKnownTypes.Get(WellKnownType.Microsoft_AspNetCore_Mvc_FromServicesAttribute);
+            if (handlerDelegateParameter.HasAttribute(fromServicesAttributeTypeSymbol))
+            {
+                continue;
+            }
+
+            var parameterTypeSymbol = ResovleParameterTypeSymbol(handlerDelegateParameter);
+
+            // If the parameter is one of the special request delegate types we can skip it.
+            if (wellKnownTypes.IsType(parameterTypeSymbol, RouteWellKnownTypes.ParameterSpecialTypes))
+            {
+                continue;
+            }
+
+            var syntax = (ParameterSyntax)handlerDelegateParameter.DeclaringSyntaxReferences[0].GetSyntax(context.CancellationToken);
+            var location = syntax.GetLocation();
+
+            // Match handler parameter against route parameters. If it is a route parameter it needs to be parsable/bindable in some fashion.
+            if (parser.TryGetRouteParameter(handlerDelegateParameter.Name, out var routeParameter))
+            {
+                if (!(ParsabilityHelper.IsTypeParsable(parameterTypeSymbol, wellKnownTypes) || ParsabilityHelper.IsTypeBindable(parameterTypeSymbol, wellKnownTypes)))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        DiagnosticDescriptors.RouteParameterComplexTypeIsNotParsable,
+                        location,
+                        routeParameter.Name,
+                        parameterTypeSymbol.Name
+                        ));
+                }
+
+                continue;
+            }
+
+            if (ReportFromAttributeDiagnostic(
+                context,
+                WellKnownType.Microsoft_AspNetCore_Mvc_FromHeaderAttribute,
```

Comment:
These shouldn't check for the attribute directly. What matters is `FromHeaderAttribute` implements `IFromHeaderMetadata`.

Check out where `Microsoft_AspNetCore_Http_Metadata_IFromHeaderMetadata` and friends is used.